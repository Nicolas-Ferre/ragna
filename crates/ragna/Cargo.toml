[package]
name = "ragna"
description = "Library for easily creating GPU-native applications"
readme = "../../README.md"
keywords = ["gpu", "rendering", "compute", "shader"]
categories = ["game-development", "graphics", "rendering", "concurrency"]
exclude = [".github", "README.md"]
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true
rust-version.workspace = true

[dependencies]
bitfield-struct.workspace = true
derive-where.workspace = true
futures.workspace = true
fxhash.workspace = true
itertools.workspace = true
once_cell.workspace = true
wgpu.workspace = true
ragna_derive.workspace = true

[lints]
workspace = true
