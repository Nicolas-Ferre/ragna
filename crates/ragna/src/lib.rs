//! A library for creating easily a GPU-native application.

mod app;
mod operations;
mod operators;
mod runner;
mod types;
mod wgsl;

pub use app::*;
pub use operators::*;
pub use types::*;

/// Transforms a Rust module to a GPU module.
pub use ragna_derive::gpu;

// TODO: adapt compilation tests
