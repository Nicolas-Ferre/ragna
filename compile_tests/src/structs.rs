fn main() {}

macro_rules! nothing {
    () => {};
}

#[ragna::gpu]
mod gpu {
    use ragna::U32;

    struct Unit;

    struct Empty {}

    struct Tuple(U32);

    struct WithLifetime<'a> {
        field: U32,
    }

    impl WithLifetime<'_> {
        nothing!();
    }

    trait MyTrait {
        nothing!();
    }
}
