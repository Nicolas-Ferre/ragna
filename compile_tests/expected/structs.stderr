error: unsupported empty struct
  --> compile_tests/src/structs.rs:11:5
   |
11 |     struct Unit;
   |     ^^^^^^

error: unsupported empty struct
  --> compile_tests/src/structs.rs:13:5
   |
13 |     struct Empty {}
   |     ^^^^^^

error: unsupported tuple struct
  --> compile_tests/src/structs.rs:15:5
   |
15 |     struct Tuple(U32);
   |     ^^^^^^

error: unsupported lifetime
  --> compile_tests/src/structs.rs:17:25
   |
17 |     struct WithLifetime<'a> {
   |                         ^^

error: unsupported item
  --> compile_tests/src/structs.rs:22:9
   |
22 |         nothing!();
   |         ^^^^^^^

error[E0106]: missing lifetime specifier
  --> compile_tests/src/structs.rs:17:5
   |
17 |     struct WithLifetime<'a> {
   |     ^ expected named lifetime parameter
   |
help: consider using the `'a` lifetime
   |
17 |     s'a, truct WithLifetime<'a> {
   |      +++

error[E0412]: cannot find type `a` in this scope
  --> compile_tests/src/structs.rs:17:25
   |
17 |     struct WithLifetime<'a> {
   |                         ^^ not found in this scope



error[E0392]: lifetime parameter `'a` is never used
  --> compile_tests/src/structs.rs:17:25
   |
17 |     struct WithLifetime<'a> {
   |                         ^^ unused lifetime parameter
   |
   = help: consider removing `'a`, referring to it in a field, or using a marker such as `PhantomData`

error[E0107]: struct takes 0 generic arguments but 1 generic argument was supplied
  --> compile_tests/src/structs.rs:17:12
   |
17 |     struct WithLifetime<'a> {
   |            ^^^^^^^^^^^^ -- help: remove the unnecessary generic argument
   |            |
   |            expected 0 generic arguments
   |

error[E0478]: lifetime bound not satisfied
  --> compile_tests/src/structs.rs:17:5
   |
17 |     struct WithLifetime<'a> {
   |     ^^^^^^
   |

error[E0495]: cannot infer an appropriate lifetime for lifetime parameter `'a` due to conflicting requirements
  --> compile_tests/src/structs.rs:17:12
   |
17 |     struct WithLifetime<'a> {
   |            ^^^^^^^^^^^^^^^^
   |

error: lifetime may not live long enough
  --> compile_tests/src/structs.rs:17:5
   |
17 |     struct WithLifetime<'a> {
   |     ^^^^^^              -- lifetime `'a` defined here
   |     |
   |     requires that `'a` must outlive `'static`